testing/helpers/qunitExtensions.js

```diff
@@ -153,6 +153,82 @@
 
 }();
 
+(function setupShadowDomMode() {
+    function getRoot() {
+        return document.querySelector('#qunit-fixture').shadowRoot;
+    }
+
+    function get(selector) {
+        return getRoot().querySelector(selector);
+    }
+
+    function isInShadowDomMode() {
+        return QUnit.urlParams['shadowDom'] && QUnit.urlParams['nojquery'];
+    }
+
+    function addShadowRootTree() {
+        const root = document.querySelector('#qunit-fixture');
+
+        if(!root.shadowRoot) {
+            root.attachShadow({ mode: 'open' });
+        }
+
+        const shadowContainer = document.createElement('div');
+        shadowContainer.part = 'shadow';
+
+        root.shadowRoot.appendChild(shadowContainer);
+    }
+
+    function clearShadowRootTree() {
+        const container = get('div');
+
+        jQuery(container).empty();
+        jQuery(container).remove();
+    }
+
+    let jQueryInit;
+
+    QUnit.testStart(function() {
+        if(!isInShadowDomMode()) {
+            return;
+        }
+
+        addShadowRootTree();
+
+        jQueryInit = jQuery.fn.init;
+
+        jQuery.fn.init = function(s, c, r) {
+            const result = new jQueryInit(s, c, r);
+            const resultElement = result.get(0);
+
+            if(!resultElement) {
+                return new jQueryInit(get(s), c, r);
+            }
+
+            if(resultElement === getRoot().host) {
+                return new jQueryInit(get('div'), c, r);
```

Comment:
why we search for some "div", not selector? Could you please explain this conditions?

Reply:
This is an edge case when we use `$('qunit-fixture')` as a selector.
When we do not use the shadow DOM we just put markup inside the `$(qunit-fixture)` element.
However, if we use shadow DOM we need to place markup to the element Shadow Root, which is that div.

Reply:
We also need this `div` because we clear it insede the TestDone hook